src/main/scala/akka/persistence/cassandra/journal/StreamMergerActor.scala

```diff
@@ -0,0 +1,134 @@
+package akka.persistence.cassandra.journal
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+import scala.util.control.Exception._
+
+import akka.actor.{Props, Actor}
+import akka.persistence.cassandra.journal.StreamMerger._
+import akka.serialization.SerializationExtension
+import com.datastax.driver.core.Session
+import com.datastax.driver.core.exceptions.DriverException
+
+object StreamMergerActor {
+  def props(config: CassandraJournalConfig, session: Session): Props =
+    Props(new StreamMergerActor(config, session))
+
+  val name = "StreamMerger"
+}
+
+/**
+ * Merges n independent physical streams into a single logical stream.
+ * Preserves causal order per persistenceId.
+ *
+ * The intent of current functionality is to work in batches. The merger retrieves data from all
+ * streams up to a maximum of configured step. Then creates a single logical stream out of the
+ * n independent streams. There is no ordering between the independent streams and the only
+ * requirement is to maintain causal order per persistenceId. When the whole batch can be merged
+ * the stream is emitted. If not (e.g. causality violation in given batch) then the merger
+ * requests again for more data.
```

Comment:
After a first review of `StreamMerger` and `StreamMergerActor`, the approach looks reasonable to me (although I didn't go into every detail).